client/beefy/src/worker.rs

```diff
@@ -351,55 +349,171 @@ where
 
 				info!(target: "beefy", "🥩 Round #{} concluded, committed: {:?}.", round.1, signed_commitment);
 
-				if self
-					.backend
-					.append_justification(
-						BlockId::Number(block_num),
-						(
-							BEEFY_ENGINE_ID,
-							VersionedFinalityProof::V1(signed_commitment.clone()).encode(),
-						),
-					)
-					.is_err()
-				{
-					// just a trace, because until the round lifecycle is improved, we will
-					// conclude certain rounds multiple times.
-					trace!(target: "beefy", "🥩 Failed to append justification: {:?}", signed_commitment);
+				if let Err(e) = self.backend.append_justification(
+					BlockId::Number(block_num),
+					(
+						BEEFY_ENGINE_ID,
+						VersionedFinalityProof::V1(signed_commitment.clone()).encode(),
+					),
+				) {
+					trace!(target: "beefy", "🥩 Error {:?} on appending justification: {:?}", e, signed_commitment);
 				}
 				self.signed_commitment_sender
 					.notify(|| Ok::<_, ()>(signed_commitment))
 					.expect("forwards closure result; the closure always returns Ok; qed.");
 
-				self.best_beefy_block = Some(block_num);
-				if let Err(err) = self.client.hash(block_num).map(|h| {
-					if let Some(hash) = h {
-						self.beefy_best_block_sender
-							.notify(|| Ok::<_, ()>(hash))
-							.expect("forwards closure result; the closure always returns Ok; qed.");
-					}
-				}) {
-					error!(target: "beefy", "🥩 Failed to get hash for block number {}: {}",
-						block_num, err);
+				self.set_best_beefy_block(block_num);
+
+				// Vote if there's now a new vote target.
+				if let Some(target_number) = self.current_vote_target() {
+					self.do_vote(target_number);
```

Comment:
Is this needed? The only reason for not having a vote target would be if rounds is not initialized, that will happen when we get a finality notification and we'll also attempt to vote after.